query-engine/core/src/query_graph_builder/write/delete.rs

```diff
@@ -104,3 +139,21 @@ pub fn delete_many_records(
 
     Ok(())
 }
+
+/// An atomic delete is a delete performed in a single operation. It uses `DELETE ... RETURNING` or
+/// similar statement.
+/// We only perform such delete when:
+/// 1. Connector supports such operations
+/// 2. The selection set contains no relation
+fn can_use_atomic_delete(query_schema: &QuerySchema, field: &ParsedField<'_>) -> bool {
+    // TODO laplab: check that the filter does not contain any predicates on relations.
```

Comment:
This is going to be a bit unfortunate. On SQL, we have ways to render relational filters without joins, which allows us to render relational filters even for inserts/delete/updates. So you might have to add another ugly capability to avoid atomic deletes with relational filters _only_ for MongoDB.

Reply:
Done!